src/trees/mod.rs

```diff
@@ -0,0 +1,141 @@
+    /// parenthesis.
+    fn depth(&self, node: Self::NodeHandle) -> u64;
+
+    /// Returns the number of nodes in the tree.
+    fn size(&self) -> usize;
+
+    /// Returns true, if the tree has no nodes.
+    fn is_empty(&self) -> bool {
+        self.size() == 0
+    }
+}
+
+/// A trait for succinct tree data structures that support [`subtree_size`] queries.
+///
+/// [`subtree_size`]: SubtreeSize::subtree_size
+pub trait SubtreeSize: Tree {
+    /// Returns the number of nodes in the subtree rooted at the given node.
+    /// This includes the node itself, meaning the minimum subtree size is 1.
+    /// If the function is called on an invalid node handle, the result is meaningless.
+    ///
+    /// Returns `None` if the `node` has no closing parenthesis (in an unbalanced parenthesis
+    /// expression).
+    fn subtree_size(&self, node: Self::NodeHandle) -> Option<usize>;
+}
+
+/// A trait for succinct tree data structures that support [`is_ancestor`] queries.
+///
+/// [`is_ancestor`]: IsAncestor::is_ancestor
+pub trait IsAncestor: Tree {
+    /// Returns true if `ancestor` is an ancestor of the `descendant` node.
+    /// Note that a node is considered an ancestor of itself.
+    ///
+    /// Returns `None` if the parenthesis expression is unbalanced and `ancestor` does not have a
+    /// closing parenthesis.
+    fn is_ancestor(&self, ancestor: Self::NodeHandle, descendant: Self::NodeHandle)
+        -> Option<bool>;
+}
+
+/// A trait for succinct tree data structures that support level-order traversal.
+pub trait LevelTree: Tree {
+    /// Returns the `level`'th ancestor of the given node, if it exists. If the level is 0, `node`
+    /// is returned. If `node` is not a valid node handle, the result is meaningless.
+    fn level_ancestor(&self, node: Self::NodeHandle, level: u64) -> Option<Self::NodeHandle>;
+
+    /// Returns the next node in the level order traversal of the tree, if it exists.
+    fn level_next(&self, node: Self::NodeHandle) -> Option<Self::NodeHandle>;
+
+    /// Returns the previous node in the level order traversal of the tree, if it exists.
+    fn level_prev(&self, node: Self::NodeHandle) -> Option<Self::NodeHandle>;
+
+    /// Returns the leftmost node at the given level, if it exists.
+    fn level_leftmost(&self, level: u64) -> Option<Self::NodeHandle>;
+
+    /// Returns the rightmost node at the given level, if it exists.
+    fn level_rightmost(&self, level: u64) -> Option<Self::NodeHandle>;
+}
+
+/// This trait provides the functionality to build a tree by visiting its nodes in depth first
+/// search order. The caller should call [`enter_node`] for each node visited in pre-order depth-first
+/// traversal, and [`leave_node`] once the node's subtree was visited (i.e. post-order).
+///
+/// Once the full tree has been visited, the caller must call [`build`] to create an instance of the
+/// implementing tree type.
+pub trait DfsTreeBuilder {
```

Comment:
I was about to write "what does Dfs" stand for, but I realize now that 'Dfs' stands for depth-first-search order. Do we anticipate other tree builders in the future and does this therefore need to be part of the name? Otherwise `TreeBuilder` would be nicer.

Reply:
I think it makes the way you have to use the builder clearer. I don't really anticipate more builders, since any other builder would be highly specific to its use-case and no more convenient to implement inside vers than outside. But it clearly signals the purpose that you have to traverse your tree structure in DFS order, and you cannot use this builder if you cannot perform that operation with your data layout.

Reply:
Well, just as a data point; it didn't make it clearer to me and I puzzled over 'Dfs' for a moment seeing it elsewhere. I tend to assume tree builders have this behavior based on experience. For someone with no experience I'm not convinced the abbreviation 'Dfs' in the name would be helpful either - the documentation would be enough. So I'd go for the more succinct name in this case, "TreeBuilder".

Reply:
Fair enough